### ISSUE TITLE
Certain GSUB replacements not working in Noto Oriya

### ISSUE THREAD

**Issue by coldbreeze16** [googlei18n/noto-fonts#579]
![clipboard01](https://cloud.githubusercontent.com/assets/13063186/11566665/63b84e7e-9a09-11e5-8a03-771011b6443f.jpg)

Certain glyph combinations defined in 'pres' features are not working in Noto Sans Oriya. In the picture Font Forge window shows the two source glyph on right and the replacement glyph on left, but as can be seen in the MS Word document the two glyphs are not replaced. This happens with a number of characters. This is by no means isolated to Windows. Mac OS X and Ubuntu also have the same issue.

![mac](https://cloud.githubusercontent.com/assets/13063186/11566801/034d6c44-9a0a-11e5-9e3a-7ae0a4e816db.jpg)

And another issue is ଜ୍ଞ appears fine in all applications but breaks down in Adobe inDesign CS6

**Comment by coldbreeze16** [comment]
I've fixed the issue with ଜ୍ଞ in inDesign by moving that rule from 'cjct' to 'akhn' (that's where it actually belongs)

**Comment by JelleBosmaMT** [comment]
The image of the TextEdit document shows conjuncts as intended. Oriya has subscript forms of consonants which are displayed below the base consonant. Some of these are diminutive forms of the base, other are abbreviated forms or signs. There is no reason to reason to display these as ligatures if there is no need. When you have a subscript consonant in combination with a second subscript, then there is a reason. You will find that lookup 25, which is highlighted in the upper image, is a lookup called by chained context lookup 23, where ligatures of various degree of compression are created depending on the size of following subscripts.
I know that most fonts insert these ligatures regardless. But think it about it: It does not make sense why some conjuncts are squished without any need for them to be squished, while others are not squished. It are the conjuncts that are combined with other marks that have these ligatures. But that does not mean they conjuncts have to be squished if there is no need for.

**Comment by JelleBosmaMT** [comment]
![ory](https://cloud.githubusercontent.com/assets/12729011/11697456/f52e1212-9eb9-11e5-8dde-bfacc7d97a1e.png)

Here are some of the ligatures you are looking for.

The nya is the only consonant without subscript, but the ligature has to be formed before the presentation features. The cjct feature has been defined to take care of this type of consonant-consonant ligatures. I cannot think of any reason why it should not belong there, and if if it does not belong there it should not matter:  the cjct feature is listed in the development specification for Oriya (ory2), so there is no excuse to ignore the substitution.

**Comment by coldbreeze16** [comment]
What you say makes sense from a technical point of view but typefaces aren't all about technicalities, they also have to deal with aesthetics. There is a reason every Oriya font in existence combines these subscript forms of consonants into a ligature. Even proprietary encoded ASCII Oriya fonts from 1980 which were PS Type 1 fonts and had a very restricted character set included many of these as a ligatures. Why? Because they look better. With the exception of a few subscript forms, most of the subscript forms partially occlude the lower part of the base character ("squishing" as you put it). This is done in written as well as printed Oriya. If you want evidence I can post photos from even older books showing that. This is a language construct and can't simply be done away with.

**Comment by coldbreeze16** [comment]
![clipboard01](https://cloud.githubusercontent.com/assets/13063186/11698929/7629dde2-9ee7-11e5-99ca-048f8e52de7e.jpg)
This is one of the older ASCII fonts. Notice how it includes the subscript forms of a few consonants. These are much less commonly used or don't have to "squish" the base glyph. 

Despite the restricted character set, notice how the designer had to include the ligatures.

**Comment by coldbreeze16** [comment]
I know cjct is listed in ory2. But it's technically never used. 
I have studied 6 Unicode Oriya fonts (excluding Noto these are all that is available in Oriya). 5 of them include ’କ୍ଷ’ and ’ଜ୍ଞ’ in akhn feature. One includes them in akhn and cjct. 

As I have pointed out including them in cjct doesn't work on inDesign (with Adobe as one of the founders of Open Type, this should matter, no?) and certain places on Mac. Now you can argue in favor of technicalities and keep it as such or improvise to maximize compatibility.

**Comment by JelleBosmaMT** [comment]
Early in the design process we showed some variants to Oriya readers, who
preferred the subscript below the base over a squished ligature of base
with subscript. 

Fonts are about aesthetics indeed. For Oriya the aesthetic element
revolves about circles. Does squishing characters make those circles look
better? In Kalinga and Normal UI କ୍ର  is deeper than କ୍ତ. Why? Does that
make କ୍ତ prettier? No it is hideous!  All these squished ligatures in
older fonts are not done despite technical limitations, they are because
of them. You must have a squished form in order to be able to put
something below and you do not have the logic to not have them if not
needed.

**Comment by JelleBosmaMT** [comment]
I have not tested against inDesign, so this is the first I hear about it
skipping cjct.  I have tested agains 3 shapers which are OK. Adobe may be
a founder of OpenType, they have no involvement with the specifications
for Indics, and what hear, their implementations are not very solid. I can
bear in mind to add akhn in future updates. But maybe you can report a bug
to Adobe too, and it gets fixed before the next update of Noto Sans Oriya.

**Comment by coldbreeze16** [comment]
It's actually funny because Noto has the actual ligature glyphs (kaTaUIorya) which is also referred to in some GPOS rules so I thought it was a bug. I don't know which Oriya users you showed it to, I'm a native Oriya user and I find it ugly. Some of us volunteers are trying to wean the local Oriya printing industry from proprietary legacy encodings to Unicode. We have developed converters and old style keyboard layouts but the main problem is the lack of fonts. Out of the 6 Unicode fonts I mentioned only 4 are feature complete (Kalinga, Noto, Lohit and OT Jagannatha) When we showed them to various publishers, DTP operators and graphic designers, they rejected Noto because these subscript forms and asked us to make them ligatures (moste specifically ta forms: କ୍ତ, ପ୍ତ). Lohit was also rejected because it's just plain ugly. Only with two fonts we can't compete with those paid software including 20-40 fonts.

As I said I'm a native speaker, these forms are called juktakshara (joined letters) for a reason: they become one ligature. I already mentioned this "squishing" is present in handwriting too. You can google and see. Here is a sample http://hindirinny.blogspot.in/2008/03/hail-mary-handwriting.html

Also your logic of the need to squish characters to accommodate below mark seems a bit odd since they could just put in another set of below marks placed further lower and in fact most of them do it. A set of lower and slighty left/right placed below marks is much easier than including so many ligatures.

**Comment by coldbreeze16** [comment]
We have raised the issue with Adobe and also been in contact with Paul Hunt from Adobe. 

We were planning to design some more OFL licensed fonts with Noto's GSUB rules but since you decide to keep your design paradigm could you at least help us in our efforts here http://forum.fontlab.com/font-formats-and-opentype/open-type-gsub-problem/

**Comment by JelleBosmaMT** [comment]
On 10-dec.-15 05:58, "coldbreeze16" <notifications@github.com> wrote:

>We were planning to design some more OFL licensed fonts with Noto's GSUB
>rules but since you decide to keep your design paradigm could you at
>least help us in our efforts here
>http://forum.fontlab.com/font-formats-and-opentype/open-type-gsub-problem/


I would not expect somebody at the FontLab forum to answer this.

Unfortunately when you have certain top marks, they are sorted in between
the base and the below base marks. You cannot use the lookup flag to
ignore marks, because the subscript consonants are marks too and would be
ignored. But there is the option to use mark attachment classes to
selectively ignore marks. This is an additional GDEF classification. You
need a class for all below base marks. For the top marks Noto uses a class
for I vowels and ligatures, and a second class for the rest (reps,
candrabindu, reps-candrabindu). So whenever you have a lookup to do things
with below base marks, set the mark attachment class to ignore the top
marks. I think different shapers might have different ordering of marks,
but with consistent use of mark attachment classes that should not matter.


>

**Comment by coldbreeze16** [comment]
I tried that.
@below_marks=[or_ka_phalaa or_ma_phalaa];

lookup blws9 {
    lookupflag MarkAttachmentType @below_marks;
    sub or_ka.pres or_ka_phalaa by or_k_ka;

Now even the ligature or_k_ka (କ୍କ) won't form. It's acting like lookup IgnoreMarks. Only the subscript form is created but ligature isn't formed.

**Comment by JelleBosmaMT** [comment]
If there had not been a survey which revealed a preference for full
subscript, Noto would have been more what you expected. And in a sense it
has all the glyphs for an alternative, conservative, version, next to the
current set of fonts. But I have been struggling to make sense of the
Oriya render rules for a very long time. And the results of the survey
felt like an eye opener. The collection of shapes that make up the Oriya
script started to make much more sense to me.


The design you are working proofs my point to some extend. You are working
on a very blocky and regular constructed design. The idea to make the
“given” ligature k-va as small as possible, makes it smaller and lighter
as the k-ra. But if you have enough descender to fit in a heavy ra, why
not use that room for k-va too? Is not the aim of a blocky regular design
to make it regular and fit into blocks?



It looks as if you are trying to make all ligatures as small as you can
get away with. And in a heavy weight as this, that gives a very ragged
lower section. In the image the below base I-vowel is the lowest. It could
(maybe should) be significant smaller, but even then. You need to have
room below the baseline for the inevitable subscripts for which you do not
have a ligature. Why not use that room? Would h-ma not be more clear and
consistent with the k-ss-ma if was the same size and weight: and fits into
the same block of the blocky font. And would k-ss-ma not be more clear and
consistent with k-va if the subscript ma has the same size as the
subscript va?  Why takes the simple ka-halant more room than the
complicated h-ma. You can create a system of various sized elements: signs
(such as for ra and ta), half forms (full forms without headstroke, which
tend to be larger as the signs) and the diminutive forms (the largest).
Create zones where these fit into (block sizes), and stick to it. It is
all about creating patterns.
In a font like this you probably aim the make the height of the font as a
whole as compact as possible. So you need squished elements and ligatures.
But then within that restricted height you should make the ligatures as
tall as possible.


I think the Oriya ligatures are (should be) a means, not a goal. The goal
is to combine base consonants with consonant signs, half-forms, diminutive
forms and vowel signs. The goal is not to create ligatures but to put
shapes on top of each other, and there is not a lot of room to do that. So
for Noto I tried to use all the room. The diminutive forms must be large
enough to distinguish at small sizes. They must fit below the baseline: we
cannot count on having a ligature for every possible loan-word or
transcription that people may come up with. So that is the room that is
available. The diminutive forms of Oriya are the most unsuitable forms to
make ligatures with of any script I know. In light fonts with small
diminutive’s the squished ligatures are bearable. But as weights become
heavier it gets mean: the subscript wants more room than the base has to
give. So if there is a survey hat says ‘no ligature if not needed to make
room for a vowel', that opens opportunities. Rather than forcing elements
together, give them room to be clear.

**Comment by JelleBosmaMT** [comment]
![block](https://cloud.githubusercontent.com/assets/12729011/11715684/ff97ef9a-9f43-11e5-9d27-04b6eccb5ab8.png)

**Comment by JelleBosmaMT** [comment]
![descender](https://cloud.githubusercontent.com/assets/12729011/11715691/0ac819e4-9f44-11e5-81e7-0dc9c221d5e9.png)

Images send by email reply did not make it. So added separately.

**Comment by JelleBosmaMT** [comment]
I cannot answer this. I am unfamiliar with Adobe layout language, I don’t
know what it means.

 I have looked at decompiled GSUB of your font on GitHub, and from that it
makes sense that it does not work without mark attachment type. But there
is a lot of other things that are different from how I structure the table
and I haven’t figured out the details of how it works or not works.

**Comment by coldbreeze16** [comment]
Actually the glyphs you are seeing are merely placeholders, I have not designed them. I took them from another font and pasted them to test the font and fix the OpenType rules. My designs are still sitting within Illustrator or Photo Shop. 

Actually what you say makes sense, in certain cases. I think with a properly designed typeface most of the conjuncts would and should assume nearly similar heights to be more harmonic. But then the ligatures have existed for far longer. 

I think most people prefer the ligature form for whatever it is worth. If you doubt that, I can put a survey on one of the popular facebook groups and send you a link. Where did you conduct your survey?

 Btw, I'm open to using your OpenType rules if I can adapt them. What tool do you use for Writing OT rules and extracting them? I'd rather be dealing with design aspect than deal with shenanigans of OT. I'm not trying anything Adobe specific.

**Comment by JelleBosmaMT** [comment]
Because I had to make Indic fonts (adaptations of old Monotype fonts that ended up in Arial Unicode Ms) before there were any decent tools to make OTL tables, I made some myself. The specification has actually been put on GitHub recently (http://monotype.github.io/OpenType_Table_Source/otl_source.html). I assume in due course some tools too.

**Comment by coldbreeze16** [comment]
I must thank you. I had only looked at the GSUB rules of Kalinga, Nirmala and a few others but not Noto. Yesterday I did exactly that and I think I have figured out the culprit. 

After wasting more than 100 hours fooling around trying various permutations and combinations over 8 months I have finally figured out what was wrong only after analyzing Google's Noto Sans Oriya's rules.

Turns out it's the incompatibility between language tags. All Indic languages have a v1 and v2 tags like <orya> vs <ory2> . And there's one particular difference that is implemented differently on most applications and inDesign. I think this is limited to the Indic languages using subscript forms as a staple (Oriya and Malayalam)

In orya you need to have
ka + virama => ka_subscript_form;

But in ory2 you need to have (which is actually correct)
virama + ka => ka_subscript_form;

I was implementing only <ory2> but now that I have included both in VOLT and with separate <blwf> rule for each, the font is working! I had tried including both earlier but with the same rules for both so it wasn't working.

**Comment by coldbreeze16** [comment]
Btw here's what MS Typography says about including ଜ୍ଞ in <akhn> . Many include କ୍ଷ too, but that's not necessary. It's because these two are considered by some Oriya authorities as part of the alphabet rather than just conjuncts. And <akhn> has the highest priority.
https://www.microsoft.com/typography/OpenTypeDev/oriya/intro.htm

**Comment by JelleBosmaMT** [comment]
It really should not matter in which place the ଜ୍ଞ ligature is created.
The akhn feature is used for ligatures that are handled like true
characters: ligatures that would have half-forms or subscript forms
themselves. As long as that is not the case, there is no reason to
prioritize it in shaping.

**Comment by coldbreeze16** [comment]
କ୍ଷ does have subscript forms, and as I mentioned ଜ୍ଞ and କ୍ଷ are sometimes considered as part of the alphabet itself. The name akhand itself means indivisible. 

But anyway that is going beyond the scope of original bug report I filed. After checking Noto's GSUB rules it's clear that ligatures not being substituted is not by mistake but by design and is deliberate. So I'm closing this. I'll be hosting a version of my font with Noto's glyph that do use the ligatures at my github.
